### PR TITLE
Kubelet: add image ref to ImageService interfaces

### LINKED ISSUE #38833: CRI PullImage / CreateContainer race
There is a race condition for CRI as currently specified:

- `PullImage` pulls an image by `name` (not `name:digest`)
- Another process changes `name` to point to something else
- `CreateContainer` creates a container using `name`, running the incorrect image

I would also classify this as a security concern, as the race could potentially be exploited to run a malicious image.

Proposed solution:

Require `PullImage` to return a fully resolved `name:digest` so that `CreateContainer` can use an immutable reference to the image it should run.

cc @kubernetes/sig-node @yujuhong

### PR DESCRIPTION
**Pull request by feiskyer**
This PR adds image ref (digest or ID, depending on runtime) to PullImage result, and pass image ref in CreateContainer instead of image name. It also

* Adds image ref to CRI's PullImageResponse
* Updates related image puller
* Updates related testing utilities

~~One remaining issue is: it breaks some e2e tests because they [checks image repoTags](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L1941) while docker always returns digest in this PR. Should we update e2e test or continue to return repoTags in `containerStatuses.image`?~~

Fixes #38833.

### CODE REVIEW

**Comment by feiskyer** [comment]
> One remaining issue is: it breaks some e2e tests because they checks image repoTags while docker always returns digest in this PR. Should we update e2e test or continue to return repoTags in containerStatuses.image?

cc/ @euank @ecordell @yifan-gu @Random-Liu WDYT

File: pkg/kubelet/images/image_manager.go [hunk 1]
```diff
@@ -81,8 +81,9 @@ func (m *imageManager) logIt(ref *v1.ObjectReference, eventtype, event, prefix,
 	}
 }
 
-// EnsureImageExists pulls the image for the specified pod and container.
-func (m *imageManager) EnsureImageExists(pod *v1.Pod, container *v1.Container, pullSecrets []v1.Secret) (error, string) {
+// EnsureImageExists pulls the image for the specified pod and container, and returnsIt returns
```
Comment (yifan-gu) [inline pkg/kubelet/images/image_manager.go]: remove `returnsIt` ?

File: pkg/kubelet/images/image_manager.go [hunk 2]
```diff
@@ -94,51 +95,52 @@ func (m *imageManager) EnsureImageExists(pod *v1.Pod, container *v1.Container, p
...
 		msg := fmt.Sprintf("Failed to inspect image %q: %v", container.Image, err)
 		m.logIt(ref, v1.EventTypeWarning, events.FailedToInspectImage, logPrefix, msg, glog.Warning)
-		return ErrImageInspect, msg
+		return "", msg, ErrImageInspect
 	}
 
-	if !shouldPullImage(container, present) {
-		if present {
+	if !shouldPullImage(container, imageRef != "") {
+		if imageRef != "" {
```
Comment (yifan-gu) [inline pkg/kubelet/images/image_manager.go]: nit: reuse variable, e.g. `present := imageRef != ""`

File: pkg/kubelet/images/puller.go [hunk 3]
```diff
@@ -24,8 +24,13 @@ import (
 	"k8s.io/kubernetes/pkg/util/wait"
 )
 
+type imageRefWithError struct {
```
Comment (yifan-gu) [inline pkg/kubelet/images/puller.go]: nit: maybe rename to pullResult?

File: pkg/kubelet/rkt/image.go [hunk 4]
```diff
@@ -66,30 +66,36 @@ func (r *Runtime) PullImage(image kubecontainer.ImageSpec, pullSecrets []v1.Secr
...
 		glog.Errorf("Failed to fetch: %v", err)
-		return err
+		return "", err
 	}
-	return nil
+	return r.getImageID(img)
 }
 
-func (r *Runtime) IsImagePresent(image kubecontainer.ImageSpec) (bool, error) {
+func (r *Runtime) IsImagePresent(image kubecontainer.ImageSpec) (string, error) {
```
Comment (yifan-gu) [inline pkg/kubelet/rkt/image.go]: @feiskyer I think you can just call `getImageID` here.
E.g. `return r.getImageID(image.Image)`

**Comment by yifan-gu** [comment]
> Should we update e2e test or continue to return repoTags in containerStatuses.image?

@feiskyer I think we should keep returning the repoTag in that status field. The e2e test is a good example that some users might reply on this field in their code.

File: pkg/kubelet/images/image_manager.go (hunk 1 above)
Comment (feiskyer) [inline pkg/kubelet/images/image_manager.go]: ack

File: pkg/kubelet/images/image_manager.go (hunk 2 above)
Comment (feiskyer) [inline pkg/kubelet/images/image_manager.go]: ack

File: pkg/kubelet/images/puller.go (hunk 3 above)
Comment (feiskyer) [inline pkg/kubelet/images/puller.go]: ack

File: pkg/kubelet/rkt/image.go (hunk 4 above)
Comment (feiskyer) [inline pkg/kubelet/rkt/image.go]: It's a little different with `getImageID`, IsImagePresent expects no error if image not found.

**Comment by feiskyer** [comment]
> So are you saying if we give docker the image digest for container creation, then later container inspect will only return the image digest?
> Can we change this field[1] to return the repoTag? We've already inspected the image here.

Exactly. Will do it this way.

**Comment by feiskyer** [comment]
Added a commit to return repoTags instead of digest in `containerStatuses.image`. 

@resouer  @yifan-gu PTAL.

File: pkg/kubelet/dockershim/docker_container.go [hunk 5]
```diff
@@ -364,10 +364,14 @@ func (ds *dockerService) ContainerStatus(containerID string) (*runtimeapi.Contai
...
 	labels, annotations := extractLabels(r.Config.Labels)
+	imageName := r.Config.Image
+	if len(ir.RepoTags) > 0 {
+		imageName = ir.RepoTags[0]
+	}
 	return &runtimeapi.ContainerStatus{
 		Id:          &r.ID,
 		Metadata:    metadata,
-		Image:       &runtimeapi.ImageSpec{Image: &r.Config.Image},
+		Image:       &runtimeapi.ImageSpec{Image: &imageName},
```
Comment (resouer) [inline pkg/kubelet/dockershim/docker_container.go]: What's the value of `ir.RepoTags[0]`? Is it meaningful enough for a field in status?

**Review by resouer — COMMENTED** [review]
Left some comments, please take into consideration

File: pkg/kubelet/container/runtime.go [hunk 6]
```diff
@@ -146,10 +146,10 @@ type IndirectStreamingRuntime interface {
 
 type ImageService interface {
 	// PullImage pulls an image from the network to local storage using the supplied
-	// secrets if necessary.
-	PullImage(image ImageSpec, pullSecrets []v1.Secret) error
+	// secrets if necessary. It returns a reference (digest or ID) to the pulled image.
+	PullImage(image ImageSpec, pullSecrets []v1.Secret) (string, error)
 	// IsImagePresent checks whether the container image is already in the local storage.
-	IsImagePresent(image ImageSpec) (bool, error)
+	IsImagePresent(image ImageSpec) (string, error)
```
Comment (feiskyer) [inline pkg/kubelet/container/runtime.go]: bool is overlapped with string (bool is actually equals to `string!=""`). It returns `("", nil)` in current implementation if image doesn't exist.

File: pkg/kubelet/dockershim/docker_container.go (hunk 5 above)
Comment (feiskyer) [inline pkg/kubelet/dockershim/docker_container.go]: e.g. "nginx:latest" for nginx image. It is the image repo (including name) with tags.

Comment (resouer) [inline pkg/kubelet/dockershim/docker_container.go]: ok, that's enough

File: pkg/kubelet/rkt/image.go (hunk 4 above)
Comment (yifan-gu) [inline pkg/kubelet/rkt/image.go]: ack

File: pkg/kubelet/container/runtime.go (hunk 6 above)
Comment (yifan-gu) [inline pkg/kubelet/container/runtime.go]: Yeah, the name `IsImagePresent` implies it should return a boolean, maybe rename the function to `GetImageDigest` or something...

File: pkg/kubelet/images/image_manager.go [hunk 7]
```diff
@@ -94,51 +95,53 @@ func (m *imageManager) EnsureImageExists(pod *v1.Pod, container *v1.Container, p
...
+		return "", msg, ErrImagePullBackOff
 	}
 	m.logIt(ref, v1.EventTypeNormal, events.PullingImage, logPrefix, fmt.Sprintf("pulling image %q", container.Image), glog.Info)
-	errChan := make(chan error)
-	m.puller.pullImage(spec, pullSecrets, errChan)
-	if err := <-errChan; err != nil {
-		m.logIt(ref, v1.EventTypeWarning, events.FailedToPullImage, logPrefix, fmt.Sprintf("Failed to pull image %q: %v", container.Image, err), glog.Warning)
+	pullChan := make(chan pullResult)
+	m.puller.pullImage(spec, pullSecrets, pullChan)
+	imageRefWithErr := <-pullChan
```
Comment (yifan-gu) [inline pkg/kubelet/images/image_manager.go]: nit: s/imageRefWithErr/imagePullResult

**Comment by yifan-gu** [comment]
Thanks @feiskyer fpr taking care of this. 
Just a couple nits, LGTM overall.

File: pkg/kubelet/container/runtime.go (hunk 6 above)
Comment (feiskyer) [inline pkg/kubelet/container/runtime.go]: Good point. However, it's not always return digest, sometimes returns ID. I think rename it to`GetImageRef` is better.

**Comment by yifan-gu** [comment]
LGTM, leave for @resouer to apply the label.

**Comment by resouer** [comment]
LGTM from my side. Thanks!!!

**Comment by resouer** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by feiskyer** [comment]
@k8s-bot gce etcd3 e2e test this

**Pull request merged**